### DISCUSSION TITLE
Simulate proper air resistance?

### DISCUSSION THREAD

**Discussion by peter1745** [jrouwe/JoltPhysics#415] [Q&A] [answered]
Hi, I was just looking around the Jolt API and I was wondering if there's any plans for adding air resistance natively to Jolt? I tried playing around with linear and angular damping, but that doesn't really give the same effect. I realize it's probably possible for me to implement it myself, either by forking Jolt or by just calculating and applying air resistance myself.

From memory PhysX and Bullet doesn't simulate air resistance natively either, and I'm curious why that is. I may end up giving it a go myself by forking Jolt, but yeah, just curious as to why it's not natively done, or if I'm just not understanding how it's supposed to be done 🙂

**Comment by jrouwe** [comment, marked as answer]
I don't think you need to fork Jolt in order to do this. In its simplest form you can write a loop to iterate over all active bodies and apply an impulse = c * v^2 * dt to the bodies (make sure they don't flip velocity). See: https://github.com/jrouwe/JoltPhysics/blob/d5cf4634292ad9a35421ffceb42a50faf3ce5562/Jolt/Physics/Body/Body.cpp#L244-L254 (note that you can call this function as well, but if you don't want buoyancy and only drag, you're wasting some cycles)

**Reply by peter1745** [reply]
Thanks for the quick reply! I ended up using `ApplyBuoyancyImpulse` for now just for simplicity, and it does work the way I want it to, I might end up copying it and stripping out the bouyancy later on just so I don't waste cycles.
Completely forgot that Jolt had buoyancy support, do you know why most physics engines don't seem to provide any air resistance by default?
Is it just not that useful generally or is there some other reason?

**Reply by jrouwe** [reply]
My guess (and my experience) is that linear damping is sufficient in most cases. Only at high velocities there is significant difference between linear and quadratic damping and the end result is that there will be some sort of terminal velocity which you can approximate by setting a max velocity on your body. A proper air resistance model can be implemented in multiple ways and I think you can make it as simple or complex as you want (ApplyBuoyancyImpulse = simple). As air resistance can easily be implemented outside the main physics update, I'm guessing that's why the major physics engines leave this as an exercise to the user. If it is important to you, you probably know your domain well and you can probably pick the right tradeoff between computational complexity and quality.
